File: addons/kebun_iot/models/sensor_data_service.py
from odoo import api, fields, models
import random


class SensorDataService(models.AbstractModel):
    _name = "kebun_iot.sensor_data_service"
    _description = "Sensor Data Service"

    SENSOR_CONFIG = {
        "ph": {"min": 6.8, "max": 8.5, "step": 0.01, "precision": 2},
        "ec": {"min": 300, "max": 900, "step": 2, "precision": 0},
        "tds": {"min": 150, "max": 600, "step": 1, "precision": 0},
        "salinity": {"min": 0.1, "max": 3.5, "step": 0.005, "precision": 2},
        "sg": {"min": 0.995, "max": 1.03, "step": 0.0002, "precision": 4},
        "temperature": {"min": 22, "max": 32, "step": 0.02, "precision": 2},
    }

    @api.model
    def _random_step(self, step):
        return (random.random() * 2 - 1) * step

    @api.model
    def _clamp(self, value, minimum, maximum):
        return max(min(value, maximum), minimum)

    @api.model
    def get_initial_values(self):
        return {
            key: config["min"] + (config["max"] - config["min"]) / 2
            for key, config in self.SENSOR_CONFIG.items()
        }

    @api.model
    def simulate_values(self, previous_values):
        values = {}
        for key, config in self.SENSOR_CONFIG.items():
            previous = previous_values.get(key, config["min"] + (config["max"] - config["min"]) / 2)
            next_value = previous + self._random_step(config["step"])
            values[key] = self._clamp(next_value, config["min"], config["max"])
        return values

    @api.model
    def get_sensor_snapshot(self, previous_values):
        values = self.simulate_values(previous_values)
        statuses = {
            key: self._get_status(key, val)
            for key, val in values.items()
        }
        trends = {
            key: self._get_trend(previous_values.get(key), values[key])
            for key in values
        }
        return {
            "values": values,
            "statuses": statuses,
            "trends": trends,
        }

    @api.model
    def _get_status(self, key, value):
        if key == "ph":
            return "normal" if 6.8 <= value <= 8.0 else "warning" if 6.5 <= value <= 8.3 else "critical"
        if key == "ec":
            return "normal" if 300 <= value <= 900 else "warning" if 250 <= value <= 950 else "critical"
        if key == "tds":
            return "normal" if 150 <= value <= 600 else "warning" if 130 <= value <= 650 else "critical"
        if key == "salinity":
            return "normal" if 0.1 <= value <= 3.5 else "warning" if 0.05 <= value <= 3.7 else "critical"
        if key == "sg":
            return "normal" if 0.995 <= value <= 1.03 else "warning" if 0.993 <= value <= 1.032 else "critical"
        if key == "temperature":
            return "normal" if 22 <= value <= 32 else "warning" if 20 <= value <= 33 else "critical"
        return "normal"

    @api.model
    def _get_trend(self, previous, current):
        if previous is None or abs(current - previous) < 0.01:
            return "→"
        return "↑" if current > previous else "↓"

File: addons/kebun_iot/static/src/js/sensor_data_service.js
export class SensorDataService {
    constructor() {
        this.previousValues = {
            ph: 7.3,
            ec: 520,
            tds: 260,
            salinity: 1.1,
            sg: 1.012,
            temperature: 24.5,
        };
        this.history = {
            ph: [],
            ec: [],
            tds: [],
            salinity: [],
            sg: [],
            temperature: [],
        };
        this.labels = [];
        this.config = {
            ph: { min: 6.8, max: 8.5, step: 0.01, precision: 2 },
            ec: { min: 300, max: 900, step: 2, precision: 0 },
            tds: { min: 150, max: 600, step: 1, precision: 0 },
            salinity: { min: 0.1, max: 3.5, step: 0.005, precision: 2 },
            sg: { min: 0.995, max: 1.03, step: 0.0002, precision: 4 },
            temperature: { min: 22, max: 32, step: 0.02, precision: 2 },
        };
    }

    _random(min, max) {
        return min + Math.random() * (max - min);
    }

    _clamp(value, min, max) {
        return Math.min(Math.max(value, min), max);
    }

    _simulateValue(key) {
        const config = this.config[key];
        const previous = this.previousValues[key];
        const next = previous + this._random(-config.step, config.step);
        const value = this._clamp(next, config.min, config.max);
        this.previousValues[key] = value;
        return Number(value.toFixed(config.precision));
    }

    _status(key, value) {
        if (key === "ph") {
            return value >= 6.8 && value <= 8.0 ? "normal" : value >= 6.5 && value <= 8.3 ? "warning" : "critical";
        }
        if (key === "ec") {
            return value >= 300 && value <= 900 ? "normal" : value >= 250 && value <= 950 ? "warning" : "critical";
        }
        if (key === "tds") {
            return value >= 150 && value <= 600 ? "normal" : value >= 130 && value <= 650 ? "warning" : "critical";
        }
        if (key === "salinity") {
            return value >= 0.1 && value <= 3.5 ? "normal" : value >= 0.05 && value <= 3.7 ? "warning" : "critical";
        }
        if (key === "sg") {
            return value >= 0.995 && value <= 1.03 ? "normal" : value >= 0.993 && value <= 1.032 ? "warning" : "critical";
        }
        if (key === "temperature") {
            return value >= 22 && value <= 32 ? "normal" : value >= 20 && value <= 33 ? "warning" : "critical";
        }
        return "normal";
    }

    _trend(key, previous, current) {
        if (previous === undefined || Math.abs(current - previous) < 0.01) {
            return "→";
        }
        return current > previous ? "↑" : "↓";
    }

    _captureHistory(values) {
        const now = new Date();
        const label = now.toLocaleTimeString([], { hour: "2-digit", minute: "2-digit", second: "2-digit" });
        this.labels.push(label);
        if (this.labels.length > 30) {
            this.labels.shift();
        }
        Object.entries(values).forEach(([key, value]) => {
            this.history[key].push(value);
            if (this.history[key].length > 30) {
                this.history[key].shift();
            }
        });
    }

    getSnapshot() {
        const previousValues = { ...this.previousValues };
        const values = Object.fromEntries(
            Object.keys(this.config).map((key) => [key, this._simulateValue(key)])
        );
        this._captureHistory(values);
        return {
            ...values,
            statuses: Object.fromEntries(Object.entries(values).map(([key, value]) => [key, this._status(key, value)])),
            trends: Object.fromEntries(Object.entries(values).map(([key, value]) => [key, this._trend(key, previousValues[key], value)])),
            history: this.history,
            labels: this.labels,
            updated_at: new Date().toLocaleTimeString([], { hour: "2-digit", minute: "2-digit", second: "2-digit" }),
        };
    }
}
